src/get-ipfs.ts

```diff
@@ -21,125 +26,144 @@ interface IpfsPkg extends ipfs {
-    permissions.push('id')
-  }
-  if (permissions.indexOf('version') < 0) {
-    permissions.push('version')
-  }
+  if (permissions.indexOf('id') < 0) permissions.push('id')
+  if (permissions.indexOf('version') < 0) permissions.push('version')
   return permissions
 }
 
-const connectPeers = async (ipfs: ipfs, peers: string[] = []) => {
-  await Promise.all(
-    peers.map(async p => {
-      try {
-        await ipfs.swarm.connect(p)
-      } catch (err) {
-        console.log('Could not connect to peer: ', p)
-      }
-    })
-  )
-  return ipfs
-}
 
-export const ipfsIsWorking = async (ipfs?: ipfs) => {
-  if (!ipfs) {
-    return false
+
+// Default
+// =======
+
+const getIpfs = async (config: config = {}) => {
+  if (ipfsInstance) return ipfsInstance
+
+  // If not set up before, start flow.
+  config = { ...config, peers: config.peers || [] }
+  ipfsInstance = await loadWindowIpfs(config)
+
+  if (ipfsInstance) {
+    console.log('window.ipfs is available!')
+    const peers = (config.peers || []).concat(config.localPeers || [])
+    await connectPeers(ipfsInstance, peers)
+
+  } else {
+    console.log('window.ipfs is not available, loading js-ipfs...')
+    ipfsInstance = await loadJsIpfs(config)
+
   }
+
+  return ipfsInstance
+}
+
+
+export default getIpfs
+
+
+
+// 🛠
+
+
+export async function ipfsIsWorking(ipfs?: ipfs): Promise<boolean> {
+  if (!ipfs) return false
+
   try {
     const id = await ipfs.id()
-    if (id.id && id.agentVersion) {
-      return true
-    } else {
-      return false
-    }
+    return !!(id.id && id.agentVersion)
   } catch (err) {
     return false
   }
 }
 
-export const loadWindowIpfs = async (config: config): Promise<ipfs | null> => {
-  let ipfs = (window as IpfsWindow).ipfs
-  if (ipfs) {
-    if (ipfs.enable) {
-      const permissions = normalizePermissions(config.permissions)
-      ipfs = await ipfs.enable({
-        commands: permissions
-      })
-    }
-    const isWorking = await ipfsIsWorking(ipfs)
-    if (isWorking) {
-      return ipfs as ipfs
-    } else {
-      return null
-    }
-  }
-  return null
-}
 
-export const loadIpfsScript = (): Promise<void> => {
-  return new Promise((resolve, reject) => {
-    const script = document.createElement('script')
-    script.src = 'https://unpkg.com/ipfs@0.38.0/dist/index.min.js'
-    script.onload = async () => {
-      return resolve()
-    }
-    script.onerror = reject
-    document.body.appendChild(script)
-  })
+export async function loadJsIpfs(config: config): Promise<ipfs | null> {
+  const ipfsPkg = await (() => { switch (typeof config.jsIpfs) {
+    case "function":  return config.jsIpfs()
+    case "string":    return loadJsIpfsThroughScript(config.jsIpfs)
+    default:          return loadJsIpfsThroughScript(UNPKG_URL)
+  }})()
+
+  if (!ipfsPkg?.create) return null
+  const peers = (config.peers || []).concat(config.browserPeers || [])
+  const ipfs = await ipfsPkg.create({ config: { Addresses: { Swarm: peers }}})
```

Comment:
Does this work? I tried locally and got `Error: no valid addresses were provided for transports [WebSockets,WebRTCStar,Circuit]`

Reply:
Yeah, I'm getting this error as well. That's what I'm trying to figure out with the new js-ipfs. This used to work, see our Fission Drive code: https://github.com/fission-suite/drive/blob/22468f5908b7715bf516c87debf05ae36c974588/src/Javascript/ipfs.js#L32-L34

Could it be something regarding "delegated routing" which they excluded now? https://blog.ipfs.io/2020-02-13-js-ipfs-0-41/ (search for "Delegated peer and content routing")

Are we doing delegated routing with us passing the peer id in the multiaddr? (eg. `/dns4/node.fission.systems/tcp/4003/wss/ipfs/QmVLEz2SxoNiFnuyLpbXsH6SvjPTrHNMU88vCQZyhgBzgw`)

Reply:
I... don't think so? Not very familiar with that. But delegated routing seems to be "okay I have this peer id, but not the full multiaddr, can you connect me?"

Reply:
It's actually surprising to me that that worked in the Fission Drive code. I thought I remembered trying this back when originally making this lib and it not working. I either had to use `ipfs.swarm.connect()` or put the addresses in `config.Bootstrap`

Reply:
Huh weird. Ok, I'll do some more digging because this is all very confusing 😅

Reply:
Ok, I changed it back to what is was (using the `connect` function), used `node.fission.systems`, and now it works again (most likely because of your load balancer changes). I got a weird `AggregateError` before which I didn't understand at all.

I think the addresses you pass in with `create` are for [listening](https://github.com/ipfs/js-ipfs/blob/master/docs/config.md#swarm) and connect does [dialing](https://github.com/ipfs/js-ipfs/blob/master/src/core/components/swarm/connect.js).
Difference is explained here: https://docs.libp2p.io/concepts/transport/#listening-and-dialing
Confusing API 🤷‍♂

Reply:
Do you still get `AggregateError`? I'm still getting that on the first connect, but successfully connecting on subsequent attempts. (Although I think this is actually a timing thing since setting a timeout of 200ms also solves the issue). Really no clue why this could be the case...

Reply:
And I actually think the daemon listens for _any_ connection. Cause I can connect to a random node that isn't listening for my address. So yeah really not sure about the addresses you pass in with `create`

Reply:
I'm not getting that aggregate error anymore for some reason 😕 I have no clue either where this comes from. It might just be something new, because libp2p had a big refactor as well.

Are you getting an error on https://drive.fission.codes/ as well?

Reply:
hmm nope no error on https://drive.fission.codes/ 
Is that using the updated get-ipfs?

I'm still getting the `AggregateError`. I'm just building `get-ipfs` and testing using my local version of `ipfs-user-settings` :thinking:

Reply:
Nah, still the old js-ipfs. I wonder how I can replicate this error, tried in a few different browsers, but no aggregate error.

Should we implement "retry connections" system? For example, it could retry the failed connections 3 times with a delay. On the other hand, I feel like js-ipfs should be responsible for this 🤷‍♂ What do you think?

Reply:
Okay I just tried this in Chrome and it worked :man_shrugging: No clue what's going on here.

Hmm yeah I'd say we do one retry if none of the connections succeed (that's how we handle peer connect failures in the CLI).

Reply:
Ah wait I actually have a sneaking suspicion that my in-browser ipfs cache could be screwed up somehow. I'll try clearing it and see if it helps

Reply:
Okay I tried the reconnect logic and it works on both Chrome & FF for me :+1:  I just pushed it up, but feel free to change it around if it's not what you were thinking

Reply:
Awesome thanks 👍 I did notice you added a `tries: number` param, but never used it. Do you want me to implement this? Also, no timeout/delay between retries?

Reply:
Quick code edits squeezed in before team meetings come back to bite me :stuck_out_tongue: 

Yes I'll fix that haha

And the timeout doesn't seem to be necessary. It either requires multiple connect attempts or a timeout of about 200ms :man_shrugging: 

I'll add it as an additional param though in case we want to use it

Reply:
😬 😄 Ok cool. So weird these connection issues.

One last thing, I'm getting an error on safari when using the latest js-ipfs. Could you check if you get that too with ipfs-user-settings? 🙏  I'm getting a `timeoutController.clear is not a function` error. Made an issue for that on the libp2p repo, but trying to get more info.

Reply:
okay pushed the new code :+1: 

I'm on linux so unfortunately don't have Safari :grimacing: but I don't get any such error in FF/Chrome

Reply:
Ah right. Carry on! 😬 😅